privacy-protections/fingerprinting/canvas.js

```diff
@@ -37,12 +37,92 @@ function generateDataURLWithCode (codeInput, alpha = 255) {
+                return d[i] + (d[i + 1] * 10 ** 3) + (d[i + 2] * 10 ** 6) + (d[i + 3] * 10 ** 9);
+            }
+
+            function shouldIgnorePixel (d, i) {
+                if (d[i + 3] === 0) {
+                    return true;
+                }
+                return false;
+            }
+
+            const pixelData = [];
+            let randData;
+            // eslint-disable-next-line new-cap
+            const rng = new Math.seedrandom('something');
+            // Compare 10 canvases and ensure per pixel there is more than one result
+            for (let i = 0; i < 10; i++) {
+                const canvasElement = createCanvas(2000, 200);
+                const canvasContext = canvasElement.getContext('2d');
+
+                // Generate rand data for all canvases to use
+                if (i === 0) {
+                    const canvasDataBlank = canvasContext.getImageData(0, 0, canvasElement.width, canvasElement.height);
+                    for (let i = 0; i < size * 4; i += 4) {
+                        canvasDataBlank.data[i] = Math.floor(rng() * 256);
+                        canvasDataBlank.data[i + 1] = Math.floor(rng() * 256);
+                        canvasDataBlank.data[i + 2] = Math.floor(rng() * 256);
+                        canvasDataBlank.data[i + 3] = 255;
+                    }
+                    randData = canvasDataBlank;
+                }
+                canvasContext.putImageData(randData, 0, 0);
+
+                // Draw a unique pixel in top left corner
+                const ctx = canvasElement.getContext('2d');
+                ctx.fillStyle = `#f6${i}`;
+                ctx.fillRect(0, 0, 1, 1);
+
+                const canvasData = canvasContext.getImageData(0, 0, canvasElement.width, canvasElement.height);
+                // Start at pixel 1 as pixel 0 was used for making the canvas random
+                for (let j = 1; j < size; j++) {
+                    if (!shouldIgnorePixel(canvasData.data, j * 4)) {
+                        if (!(pixelData[j])) {
+                            pixelData[j] = new Set();
+                        }
+                        pixelData[j].add(checkPixel(canvasData.data, j * 4));
+                    }
+                }
+            }
+
+            const sizeCount = [];
+            for (let i = 1; i < size; i++) {
+                // The number of unique pixel data across the slices
+                const pixelSize = pixelData[i].size;
+                if (!(pixelSize in sizeCount)) {
+                    sizeCount[pixelSize] = 0;
+                }
+                sizeCount[pixelSize] += 1;
+            }
+            const zeroChange = sizeCount.shift();
+            ok(zeroChange === undefined, 'Sanity check on ignored pixels');
+            const singleChange = sizeCount.shift();
+            const changeCount = sizeCount.reduce((a, i) => a + i, 0);
+            const elegiblePixels = singleChange + changeCount;
```

Comment:
I don't think this is calculated correctly if I understand correctly what you're going for here. Eligible pixels should be a count of the total number of pixels that aren't transparent, whereas this is counting the number of unique values of those pixels. Instead, you can count eligiblePixels in the loop above that checks `shouldIgnorePixel`.

Then if you want to check that at least N% have been changed you also can't use uniques because it might be the case that some pixels were changed to the same value. To count those you can just loop through the final canvas and then increment if the pixel is different than `randData` at the same index.

Reply:
I sum together number of pixels with a single change and the number of pixels that have more than one (for clarity the sizeCount is an array that sums to the complete pixel array size minus the 0th pixel that we ignore). The number is the same but I changed to calculating it in the loop above to simplify the understanding.

I've updated the var names and comments here.